### PR TITLE
Implement lazy rollover for failure stores

### PR DESCRIPTION
**Pull request by nielsbauman**
Allow marking failure stores for lazy rollover. When a failure occurs while trying to ingest a document, and it gets redirected to a failure store that is marked for lazy rollover, the failure store will be rolled over and the document gets redirected to the new write index of the failure store. Currently, the two places where we roll over failure stores that are marked for lazy rollover are 1. `TransportBulkAction.java` (failed ingest node operations) and 2. `BulkOperation.java` (shard-level bulk failures). Note that there is currently no automation in place for marking failure stores for lazy rollover. This is available through the rollover API, but it will most likely mainly be used by a future PR that does automate the marking for lazy rollover in some way.

### CODE REVIEW

**Review by gmarouli — COMMENTED** [review]
Thanks for picking this up @nielsbauman ! I do not see why it's not a ready PR. Do you have any concerns?

**Comment by nielsbauman** [comment]
@gmarouli
> I do not see why it's not a ready PR. Do you have any concerns?

I marked it as a draft as it's based off of #107562, which was still pending your review at the time of writing this :) (and I'll have to rebase this branch off of main before I can merge).

**Review by gmarouli — CHANGES_REQUESTED** [review]
Wow, this is much more complex than the lazy rollover itself. Great work getting a crack at it. Here are some initial thoughts, and I will go over it again a bit later. 💪

File: modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/200_rollover_failure_store.yml
Comment (nielsbauman) [reply modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/200_rollover_failure_store.yml:193]: Ah yeah I agree that the checks for the error message and the error stack trace aren't of much value here. Will remove :+1:

File: server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:157]: I'm having trouble envisioning a scenario where that would protect against unwanted/erroneous behavior. Could you sketch a scenario perhaps?

Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:157]: This is protecting us in the case when a cluster is being upgraded and not all nodes support this feature yet.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:209]: My reasoning for doing it like this was that I wanted to keep the default indices options of the `RolloverRequest` and wanted to avoid specifying the default myself (i.e. `IndicesOptions.strictSingleIndexNoExpandForbidClosed()`). All these rollover requests will have the same default indices options, so I'm trying to come up with a way to avoid recreating them, but I'm having trouble coming up with an _elegant_ one. At the same time, I don't expect there to be a lot of failure stores to be rolled over in one request, don't you think?

File: server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:157]: Sorry I should have elaborated. I understand that node features are useful for such cases in general. My question was more about this specific case. If we would add this check to the `TransportRolloverAction#masterOperation`, it means that this check would only be executed on master nodes that have this feature already. Rollover/lazy rollover is only dependent on the master node, right? So, I don't see in what scenario we would be executing this check, but we can't/shouldn't perform the rollover.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:204]: Yep, that's right. Are you suggesting just renaming the `failureStore` variable to `dataStream`, or also `failureStoresToBeRolledOver` to `dataStreamsToBeRolledOver`? I don't think I'm a fan of the latter. and the former might be a bit confusing because of the switch from `failureStoresToBeRolledOver` to `dataStream`. Maybe something like `failureStoreDataStreamName`? Although that's a bit too length IMO

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:468]: Good point. I'll rename it to `maybeMarkFailureStoreForRollover`. Let me know if you prefer a different name.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:183]: I'll rename to `executeRedirectRequests`.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:386]: I wanted to avoid using `computeIfAbsent` as I don't think it's very performant when executed a lot (which is the case for the bulk API). I'll update this to use `Map.containsKey`.

File: server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:157]: One scenario I could think about is that this master has been upgraded but you do not know if the other masters have. 

The problem with these things is that it's hard to argue what would happen if a new feature is used in a partially upgraded cluster, because it's very hard to think all the possible scenarios and failovers. By not protecting this you have two problems:
1. The user things the feature is properly supported, you accepted the request so it should be fine.
2. This request changes changes the cluster state by setting a pre-existing flag to `true` but it also has a later side-effect, the actual rollover which might or might not be implemented at that node either.

Does this make sense?

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:209]: > My reasoning for doing it like this was that I wanted to keep the default indices options of the RolloverRequest and wanted to avoid specifying the default myself (i.e. IndicesOptions.strictSingleIndexNoExpandForbidClosed()). 

What about a local constant, that belongs in this class only?

> All these rollover requests will have the same default indices options, so I'm trying to come up with a way to avoid recreating them, but I'm having trouble coming up with an elegant one. At the same time, I don't expect there to be a lot of failure stores to be rolled over in one request, don't you think?

If the problem is a faulty document probably not, but if the cause of the issue is a problem in the mapping, then it's possible that all the requests will fail :). There are issues that could potentially cause all index operations to fail.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:204]: I was considering to use `dataStream` or `dataStreamName` just here. `failureStoresToBeRolledOver` can stay because as you said it will cause conflicts.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:468]: I was trying to make it follow the convention of `addDocumentToRedirectRequests` but I find your suggestion also clear, so no objection there.

File: server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:157]: Ah yeah that definitely makes sense. Thanks for the explanation!

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:386]: Sorry maybe performant isn't the right word. I was thinking of https://github.com/elastic/elasticsearch/commit/99e72b02ba46038d898e981f9c39d07a392735f8, which I remembered as "the lambda argument of `computeIfAbsent` can take up significant space on the heap when invoked often". After reading the commit description again, I see it's a bit more nuanced than that. But I'm inclined to think that that statement also applies to this situation (as this will be called for every request in every bulk request). What do you think?

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:425]: I don't think it can/should, as we also need to pass a map to `executeBulk` in the optimized path, so we might as well use this instance, right?

Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:428]: Isn't it an empty immutable map then? If yes, it could use the constant `Map.of()`, otherwise you are right, we can ignore this comment.

Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:428]: Yep that's right. I thought the difference between an empty `HashMap` and an empty immutable map would be minimal but I see now that it's valuable to use an empty immutable map when possible. Thanks

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (jbaiera) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:400]: Writes to the failure store are only ever index operations. Deletes are not considered, and updates should be disallowed on data streams. I think it's safe to make this assumption here for writing to the failure store.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (jbaiera) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:391]: Is this correct? If we have any request in the bulk that requires alias, then in the previous logic we would not add the index to the auto creation. The new logic looks like as long as a single operation does not require an alias it will be added to the auto creation.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (jbaiera) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:234]: This loop makes me uneasy. In practice it's probably closer to O(n) where n is the number of documents in the bulk operation, but technically it is O(n^2) in the worst case, which would be if each document was destined to a different failure store and each rollover failed.

It looks like this is how things currently work in the bulk operation for lazy rollover, and I don't know how frequently we see lazy rollover failures, so you'd need to torture your cluster in a unique way to get this to run n^2, but it's possible. Maybe we should file a follow up to collect all the failed rollovers and traverse the documents one time using a set of failed rollovers to update the response. We'd need to fix it here and in the transport action.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (jbaiera) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:191]: Is there a chance we could end up with a new cluster block here? I don't know if we should do another round of ensuring blocks are handled or not.

**Review by jbaiera — CHANGES_REQUESTED** [review]
This is indeed a complicated one, but it looks like a good set of changes so far. I left a couple of requests for changes and a couple of comments/questions as well.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:400]: > Deletes are not considered

In my current logic (`writeToFailureStore == false && dataStream.getBackingIndices().isRolloverOnWrite()`), a delete request to a data stream that is marked for lazy rollover (i.e. that backing indices are marked for rollover, not the failure store) will roll over the data stream. My TODO question was directed toward that specific scenario.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:387]: I'm okay with extracting the lambda to outside the loop. That would still create a new lambda for every bulk request, but that's fine, I think -- extracting it to a class method or a static class field is overkill IMO. I think readability is still fine with this approach.

Also, technically, I guess `computeIfAbsent` still makes use of a tri-state boolean, it's just hidden away from us :)

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:395]: I don't think I can extract a lambda to outside the loop in this case. Can you explain what you don't like about "tri-state booleans"? I think it's a pretty common pattern to use an `== null` check to determine if a key is present in a map (as it avoids an extra `containsKey` call). Do you have any suggestions that don't require creating a new lambda inside the loop?

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:191]: I'm not too familiar with cluster blocks, but wouldn't it always be possible to get a cluster block right after obtaining a state that doesn't have one? Our `ClusterStateObserver` waiting for and obtaining a cluster state that doesn't have a block, doesn't mean that the cluster can't get blocked right after (when we start grouping the request by shard or start sending rollover requests), right?

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Ah I think you're right, good catch! Any thoughts on a good solution to this? The best solution I can think of is what I did in [3b6a10a](https://github.com/elastic/elasticsearch/pull/108108/commits/3b6a10a18293eb680b3b9ea2b9374b200e8bd350), but it requires yet another `Set` allocation and makes the loop more complex...

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:216]: Good point, still though a template change could affect a lot of data streams. In any case. You can leave it as is. No strong opinion.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:513]: Already discussed in a different comment.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:422]: That scenario is not possible.

Using the data stream name you cannot delete a document: `only write ops with an op_type of create are allowed in data streams`. 

Using the backing index name you can delete the document but that will not trigger a rollover because the `rolloverOnWrite` is on data stream level.

Does this answer your TODO?

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:191]: I think this method just returns the last observed value right? It seems to me that it just retrieves the latest one.

https://github.com/elastic/elasticsearch/blob/08c7b655a2dff7f0c66db5247a7f7fd0dd66fa88/server/src/main/java/org/elasticsearch/cluster/ClusterStateObserver.java#L92-L99

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:422]: Yep definitely, thanks both!

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:191]: Yeah I agree with what @jbaiera was saying. My point was that I don't really know what the value would be of doing another check, as even if we do a check, the cluster could become blocked 1ms after we do the check, and we'd proceed rolling over the lazy rollover, right?

Comment (jbaiera) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:191]: > the cluster could become blocked 1ms after we do the check, and we'd proceed rolling over the lazy rollover

This is the reason I'm kind of on the fence about doing anything about it. I wanted to bring it up just to make sure we were aware about it, but after taking some time to dig through the cluster blocks some more I don't think I have a preference for making a change here at the moment.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (jbaiera) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: I think if I were doing this from scratch I would have added some fields to `ReducedRequestInfo` to track if an index request was being executed regularly and if it was a failure store operation. Then I'd use the merge function on the map collector beforehand to OR the flags together. When it would come time to generate the rollovers I'd have generated a rollover request for each set flag on the `ReducedRequestInfo` for an index.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:191]: Thanks for bringing it up and doing some digging, @jbaiera!

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Yeah I thought about making changes to `ReducedRequestInfo` initially as well, but because that's an enum and not a record, I'd have to create all kinds of permutations which seemed confusing to me.

Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Are you ok with the current implementation? Or do you want me to take another crack at it? I've spent some time weighing off different approaches already, so not sure how far I'd come.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: I am ok with the `contains` too I was just as surprised as @jbaiera that's why I asked.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: Usually a `boolean` has two values, true or false. When using `Boolean` you also have the option to have `null`. So there are 3 possible values, sometimes, null just defaults to true or false, but in this case it means something else, it means what `null` usually means, I do not have a value for this and there is a different action you need to take. 

I can see how using the `contains` does not help with readability either because without `requiresDataStream = indicesToAutoCreate.get(request.index())` it's hard to remember what does that boolean represent.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java
Comment (jbaiera) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: I think your latest revision is likely the best approach for this. Good stuff!

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamFeatures.java
Comment (jbaiera) [inline modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamFeatures.java:41]: Do we need to add a node feature for now? We still have a way to limit rest test versions on the old version schemes until 8.16 is branched. If this feature wasn't behind a feature flag this would certainly be needed, but since it is do we need this?

I would propose that we use the feature flag check to guard the new logic instead of a node feature. When we remove the feature flag we can introduce a single node feature for the premiere of failure stores. Thoughts?

Comment (nielsbauman) [reply modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamFeatures.java:41]: I added the feature as per Mary's [suggestion](https://github.com/elastic/elasticsearch/pull/108108#discussion_r1599868090) - not for YAML REST tests, but for validation in [TransportRolloverAction.java](https://github.com/elastic/elasticsearch/pull/108108/files#diff-f165f6db589dbdcbde3400f8377579f9025c4173a192df11d35c9d8fd80ad703) and [BulkOperation.java](https://github.com/elastic/elasticsearch/pull/108108/files#diff-318211e7d42c8640831a65881e33020a927a2ce6342cf744bdad30ca755f9aed). Or are you suggesting to use the feature flag in those checks?

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:234]: Filed https://github.com/elastic/elasticsearch/issues/108838

File: modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamFeatures.java
Comment (jbaiera) [reply modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamFeatures.java:41]: I think we should just use the feature flag since we don't expect people to be doing rolling upgrades with feature flags enabled. 

We'll still need to guard against the test cases running for older versions. I think we should be able to set up something using the new capabilities API - preferably a test specific capability on the roll over rest API that we can remove/replace when we take the feature into tech preview.

Comment (nielsbauman) [reply modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamFeatures.java]: See [06c8b26](https://github.com/elastic/elasticsearch/pull/108108/commits/06c8b26546a87d3cb80ca543c3866e969b90a05e)

File: modules/data-streams/build.gradle
Comment (mark-vieira) [inline modules/data-streams/build.gradle:17]: @thecoop @rjernst I wonder if we should add this by default for all YAML test projects as I assume it's necessary if you use any capabilities-based `requires` conditions?

**Review by mark-vieira — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java
Comment (jbaiera) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestRolloverIndexAction.java:52]: So, I think we should only expose this conditionally as part of the feature flag being on, otherwise we add a capability to the general release that is not actually supported yet.

```suggestion
        if (DataStream.isFailureStoreFeatureFlagEnabled()) {
            return Set.of("lazy-rollover-failure-store");
        } else {
            return Set.of();
        }
```

I'd appreciate someone poking holes in this idea though. I'm engaging with the core infra folks on how best to approach using these new API's with feature flags. Unfortunately, we've mostly spun our wheels on coming up with a best practice. Right now, I think conditionally registering the capability to the rest end point is probably fine to start, but when we remove the feature flag we'll need to go through and make sure features/capabilities are correctly tagged up.

**Review by jbaiera — APPROVED** [review]
LGTM pending final comments and green CI

Thanks for iterating on this one Niels, it's really a foundational change for failure store!

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:91]: Yes, that is correct.

**Review by gmarouli — APPROVED** [review]
LGTM! Great work on tough one, well done!

**Pull request merged**
